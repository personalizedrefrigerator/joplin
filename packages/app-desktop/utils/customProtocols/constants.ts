import { dirname } from 'path';

// eslint-disable-next-line import/prefer-default-export
export const contentProtocolName = 'joplin-content';
export const appDir = dirname(dirname(__dirname));
