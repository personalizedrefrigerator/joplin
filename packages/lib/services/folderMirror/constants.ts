import { ModelType } from '../../BaseModel';
import { FolderEntity, NoteEntity, ResourceEntity } from '../database/types';
import { FolderItem } from './types';
const { ALL_NOTES_FILTER_ID } = require('../../reserved-ids.js');

export const resourcesDirName = 'resources';
export const resourceMetadataExtension = '.metadata.yml';
export const resourcesDirId = 'resource1resource1resource111111';
export const resourcesDirItem: FolderItem = {
	id: resourcesDirId,
	title: 'resources',
	parent_id: '',
	deleted_time: 0,
	virtual: true,
	type_: ModelType.Folder,
};

export const baseItem: FolderEntity = {
	id: ALL_NOTES_FILTER_ID,
	title: '',
	type_: ModelType.Folder,
};

type ItemKey = (keyof NoteEntity)|(keyof FolderEntity)|(keyof ResourceEntity);

// These fields will be compared to determine whether an item has changed with respect to a remote item.
export const itemDiffFields: ItemKey[] = [
	'title', 'body', 'icon', 'blob_updated_time', 'is_todo', 'todo_completed',
];

// When a conflict occurs, all other fields are merged by selecting the version in the
// last-updated note.
export const nonMergableFields: ItemKey[] = [ 'body', 'blob_updated_time' ];
