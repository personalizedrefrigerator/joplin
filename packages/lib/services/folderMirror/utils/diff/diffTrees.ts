import ItemTree from '../../ItemTree';
import { AddItemCommand, TreeCommand, TreeCommandType } from './commands';
import itemsMatch from '../itemsMatch';

type PathString = string;
export type TreeDiff = Map<PathString, TreeCommand>;
type IdString = string;

// Returns operations that can be applied to `originalTree` to get `updatedTree`.
const diffTrees = (originalTree: ItemTree, updatedTree: ItemTree): TreeDiff => {
	// Maps from id -> command
	const additions: Map<IdString, AddItemCommand> = new Map();
	const output: TreeDiff = new Map();

	const handledPaths = new Set<string>();

	for (const [itemPath, item] of updatedTree.items()) {
		if (handledPaths.has(itemPath)) continue;

		// Focus on diffing with paths, rather than IDs. The ID property can be unreliable when,
		// for example, a file is duplicated in the remote tree.
		if (!originalTree.hasPath(itemPath)) {
			additions.set(item.id, {
				type: TreeCommandType.Add,
				path: itemPath,
				item,
				itemType: item.type_,
			});
		} else {
			const originalItem = originalTree.getAtPath(itemPath);
			if (!itemsMatch(originalItem, item)) {
				output.set(itemPath, {
					type: TreeCommandType.Update,
					path: itemPath,
					// TODO: Handle the case where the ID changes?
					newItem: item,
					itemType: item.type_,
				});
			}
		}
		handledPaths.add(itemPath);
	}

	// Check for deletions
	for (const [path, remoteItem] of originalTree.items()) {
		if (handledPaths.has(path)) continue;

		// Item with the removed item's ID created? Move/rename.
		if (remoteItem.id && additions.has(remoteItem.id)) {
			const addCommand = additions.get(remoteItem.id);
			additions.delete(remoteItem.id);

			output.set(path, {
				type: TreeCommandType.Move,
				originalPath: path,
				newPath: addCommand.path,
				itemType: addCommand.item.type_,
			});
		} else {
			output.set(path, {
				type: TreeCommandType.Remove,
				path,
				itemType: remoteItem.type_,
			});
		}
	}

	for (const command of additions.values()) {
		output.set(command.path, command);
	}
	return output;
};

export default diffTrees;
