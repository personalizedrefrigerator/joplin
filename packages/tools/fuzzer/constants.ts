import { dirname, join } from 'path';

export const packagesDir = dirname(dirname(__dirname));
export const cliDirectory = join(packagesDir, 'app-cli');

