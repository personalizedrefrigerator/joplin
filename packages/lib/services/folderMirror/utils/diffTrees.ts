import { itemDiffFields } from '../constants';
import keysMatch from './keysMatch';
import ItemTree from '../ItemTree';
import { AddItemCommand, TreeCommand, TreeCommandType } from '../model/commands';

// Returns operations that can be applied to `originalTree` to get `updatedTree`.
const diffTrees = (originalTree: ItemTree, updatedTree: ItemTree) => {
	// Maps from id -> command
	const additions: Map<string, AddItemCommand> = new Map();

	const updates: TreeCommand[] = [];
	const handledPaths = new Set<string>();

	for (const [itemPath, item] of updatedTree.items()) {
		if (handledPaths.has(itemPath)) continue;

		// Focus on diffing with paths, rather than IDs. The ID property can be unreliable when,
		// for example, a file is duplicated in the remote tree.
		if (!originalTree.hasPath(itemPath)) {
			additions.set(item.id, {
				type: TreeCommandType.Add,
				path: itemPath,
				item,
				itemType: item.type_,
			});
		} else {
			const originalItem = originalTree.getAtPath(itemPath);
			if (!keysMatch(originalItem, item, itemDiffFields)) {
				updates.push({
					type: TreeCommandType.Update,
					path: itemPath,
					// TODO: Handle the case where the ID changes?
					newItem: item,
					itemType: item.type_,
				});
			}
		}
		handledPaths.add(itemPath);
	}

	const output: TreeCommand[] = [];

	// Check for deletions
	for (const [path, remoteItem] of originalTree.items()) {
		if (handledPaths.has(path)) continue;

		// Item with the removed item's ID created? Move/rename.
		if (additions.has(remoteItem.id)) {
			const addCommand = additions.get(remoteItem.id);
			additions.delete(remoteItem.id);

			output.push({
				type: TreeCommandType.Move,
				originalPath: path,
				newPath: addCommand.path,
				itemType: addCommand.item.type_,
			});
		} else {
			output.push({
				type: TreeCommandType.Remove,
				path,
				itemType: remoteItem.type_,
			});
		}
	}

	for (const command of additions.values()) {
		output.push(command);
	}
	return output.concat(updates);
};

export default diffTrees;
