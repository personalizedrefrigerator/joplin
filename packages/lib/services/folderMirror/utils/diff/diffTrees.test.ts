import { ModelType } from '../../../../BaseModel';
import { baseItem } from '../../constants';
import ItemTree, { noOpActionListeners } from '../../ItemTree';
import diffTrees from './diffTrees';

describe('folderMirror/diffTrees', () => {
	test('should diff with the empty tree', async () => {
		const emptyTree = new ItemTree(baseItem);
		const updatedTree = new ItemTree(baseItem);
		await updatedTree.addItemTo('', { title: 'Test', id: '12345678901234567890123456789012', type_: ModelType.Note }, noOpActionListeners);

		expect(diffTrees(emptyTree, updatedTree)).toMatchSnapshot('diff: empty vs single note');

		await updatedTree.addItemTo('', { title: 'Note 2', id: '22345678901234567890123456789012', type_: ModelType.Note }, noOpActionListeners);
		await updatedTree.addItemTo('', { title: 'Folder 1', id: '32345678901234567890123456789012', type_: ModelType.Folder }, noOpActionListeners);
		await updatedTree.addItemTo('Folder 1', { title: 'Test', id: '22345678901234567890123456789013', type_: ModelType.Note }, noOpActionListeners);

		expect(diffTrees(emptyTree, updatedTree)).toMatchSnapshot('diff: empty vs folder and notes');
	});

	test('diffs should include updates to items', async () => {
		const baseTree = new ItemTree(baseItem);
		const updatedTree = new ItemTree(baseItem);
		const testNote = { title: 'Test', id: '12345678901234567890123456789012', body: '', type_: ModelType.Note };
		await baseTree.addItemTo('', testNote, noOpActionListeners);
		await updatedTree.addItemTo('', { ...testNote, body: 'updated!' }, noOpActionListeners);

		expect(diffTrees(baseTree, updatedTree)).toMatchSnapshot();
	});

	test('diffs should include deletions', async () => {
		const baseTree = new ItemTree(baseItem);
		const updatedTree = new ItemTree(baseItem);
		const testNote = { title: 'Test', id: '12345678901234567890123456789012', body: '', type_: ModelType.Note };
		await baseTree.addItemTo('', testNote, noOpActionListeners);

		expect(diffTrees(baseTree, updatedTree)).toMatchSnapshot();
	});
});
