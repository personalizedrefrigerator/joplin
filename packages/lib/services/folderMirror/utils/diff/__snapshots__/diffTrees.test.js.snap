// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`folderMirror/diffTrees diffs should include deletions 1`] = `
Map {
  "Test.md" => {
    "itemType": 1,
    "path": "Test.md",
    "type": "remove",
  },
}
`;

exports[`folderMirror/diffTrees diffs should include updates to items 1`] = `
Map {
  "Test.md" => {
    "itemType": 1,
    "newItem": {
      "body": "updated!",
      "id": "12345678901234567890123456789012",
      "parent_id": "c3176726992c11e9ac940492261af972",
      "title": "Test",
      "type_": 1,
    },
    "path": "Test.md",
    "type": "update",
  },
}
`;

exports[`folderMirror/diffTrees should diff with the empty tree: diff: empty vs folder and notes 1`] = `
Map {
  "Test.md" => {
    "item": {
      "id": "12345678901234567890123456789012",
      "parent_id": "c3176726992c11e9ac940492261af972",
      "title": "Test",
      "type_": 1,
    },
    "itemType": 1,
    "path": "Test.md",
    "type": "add",
  },
  "Note 2.md" => {
    "item": {
      "id": "22345678901234567890123456789012",
      "parent_id": "c3176726992c11e9ac940492261af972",
      "title": "Note 2",
      "type_": 1,
    },
    "itemType": 1,
    "path": "Note 2.md",
    "type": "add",
  },
  "Folder 1" => {
    "item": {
      "id": "32345678901234567890123456789012",
      "parent_id": "c3176726992c11e9ac940492261af972",
      "title": "Folder 1",
      "type_": 2,
    },
    "itemType": 2,
    "path": "Folder 1",
    "type": "add",
  },
  "Folder 1/Test.md" => {
    "item": {
      "id": "22345678901234567890123456789013",
      "parent_id": "32345678901234567890123456789012",
      "title": "Test",
      "type_": 1,
    },
    "itemType": 1,
    "path": "Folder 1/Test.md",
    "type": "add",
  },
}
`;

exports[`folderMirror/diffTrees should diff with the empty tree: diff: empty vs single note 1`] = `
Map {
  "Test.md" => {
    "item": {
      "id": "12345678901234567890123456789012",
      "parent_id": "c3176726992c11e9ac940492261af972",
      "title": "Test",
      "type_": 1,
    },
    "itemType": 1,
    "path": "Test.md",
    "type": "add",
  },
}
`;
